main/server.go

```diff
@@ -0,0 +1,313 @@
+// CONDITIONS OF ANY KIND, either express or implied.
+//
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build appenginevm
+// This package implements a simple HTTP server providing a REST API to a todo
+// handler.
+//
+// It provides six methods:
+//
+// 	GET	/todos		Retrieves all the todos.
+// 	POST	/todos		Creates a new todo given a title.
+//	DELETE	/todos		Deletes all completed todos.
+// 	GET	/todos/{todoID}	Retrieves the todo with the given id.
+// 	PUT	/todos/{todoID}	Updates the todo with the given id.
+//	DELETE	/todos/{todoID}	Deletes the todo with the given id.
+//
+// Every method below gives more information about every API call, its
+// parameters, and its results.
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+
+	"github.com/GoogleCloudPlatform/gcloud-golang-todos/todo"
+	"github.com/gorilla/mux"
+	"golang.org/x/net/context"
+	"google.golang.org/appengine"
+	"google.golang.org/appengine/log"
+)
+
+const PathPrefix = "/api/todos"
+
+// learnJson contains marhsalled JSON served in place of TodoMVC's standard
+// learn.json and represents a merged JSON object containing both learn.json
+// and backend_learn.json.
+// TodoMVC's client-side contains some logic which renders it prettily in
+// the sidebar.
+var learnJson []byte
+
+func init() {
+	r := mux.NewRouter()
+	r.HandleFunc(PathPrefix,
+		errorHandler(ListTodos)).Methods("GET")
+	r.HandleFunc(PathPrefix,
+		errorHandler(NewTodo)).Methods("POST")
+	r.HandleFunc(PathPrefix,
+		errorHandler(DeleteCompletedTodos)).Methods("DELETE")
+	r.HandleFunc(PathPrefix+"/{id}",
+		errorHandler(GetTodo)).Methods("GET")
+	r.HandleFunc(PathPrefix+"/{id}",
+		errorHandler(UpdateTodo)).Methods("PUT")
+	r.HandleFunc(PathPrefix+"/{id}",
+		errorHandler(DeleteTodo)).Methods("DELETE")
+	http.Handle("/", r)
+	http.HandleFunc("/learn.json", RenderSidebarJson)
+	http.HandleFunc("/api", IsApiEnabled)
+}
+
+// badRequest should be returned by a handler function to indicate a malformed request.
+type badRequest struct{ error }
+
+// notFound should be returned by a handler function to indicate that the requested data could not be found.
+type notFound struct{ error }
+
+// IsApiEnabled writes an HTTP 200 indicating that the TodoMvc API is enabled for this app.
+func IsApiEnabled(w http.ResponseWriter, r *http.Request) {
+	w.WriteHeader(http.StatusOK)
+}
+
+// errorHandler wraps a function returning an error by handling the error and returning a http.Handler.
+// If the error is of the one of the types defined above, it is handled as described for every type.
+// If the error is of another type, it is considered as an internal error and its message is logged.
+func errorHandler(f func(w http.ResponseWriter, r *http.Request, c context.Context) error) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		c := appengine.NewContext(r)
+		err := f(w, r, c)
```

Comment:
I've got the tests that come with todomvc-api, though I haven't added them to the repo since it feels like cruft.

I'd really like tests written in Go, but the aetest package hasn't been ported yet.